### PR TITLE
Close #5 integration tests

### LINKED ISSUE #5: Integration tests
Set up integration test infrastructure. 

Use https://medium.com/entria/testing-a-graphql-server-using-jest-4e00d0e4980e as an inspiration or propose an alternative, everything goes.

Definition of done:
- Service layer dependencies are properly mocked out
- Integration test suite is run with _Jest_
- Integration test suite can be run separately from unit tests
- Expectations are performed with [snapshots](https://facebook.github.io/jest/docs/snapshot-testing.html)

### PR DESCRIPTION
**Pull request by aeldar**
Close #5 integration tests.
Added nock - HTTP mocking and expectations library.
Added example mock for joke service.
Added example tests for joke branch on graphql schema.
Added example test using jest's snapshots.
Jest run scripts refactored to use --project feature.

### CODE REVIEW

**Comment by aeldar** [comment]
Currently all URI are being mocked in api-mock.ts file (using nock). In fact, this file is run before every test. I didn't find a way to run it once before all tests. That is why this way of mocking web api could be not very good for performance. Another way to start mocks once is to run a separate express server in a special npm task (as proposed @trioletas . Probably this is a future improvement task.

File: package.json [hunk 1]
```diff
@@ -38,6 +40,7 @@
     "chokidar-cli": "1.2.0",
     "jest": "20.0.1",
     "jest-junit": "1.5.1",
+    "nock": "9.0.13",
```
Comment (trioletas) [inline package.json]: Please add `@types/nock` too

File: src/graphql/jest.config.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  "moduleFileExtensions": [
+    "ts",
+    "js"
+  ],
+  "roots": [
+    "./tests/"
+  ],
+  "setupFiles": [
+    "./tests/__setup/api-mock.ts"
```
Comment (trioletas) [inline src/graphql/jest.config.js]: same goes for setup => `__setup__`

File: src/graphql/tests/joke.test.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    joke {
+      text
+    }
+  }
+`;
+const rootValue = {};
+const context = {};
+
+describe('joke schema', () => {
+  it('should answer with a joke text containing \'Chuck Norris\' words', async () => {
```
Comment (trioletas) [inline src/graphql/tests/joke.test.ts]: this scenario is redundant if we do snapshot match already, leave the content assertions to the snapshot engine

File: src/graphql/tests/joke.test.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    const result = await graphql(schema, query, rootValue, context);
+    const { data } = result;
+
+    expect(data.joke).toBeTruthy();
+    expect(data.joke.text.includes('Chuck Norris')).toBe(true);
+  });
+
+  it('should answer with a joke as stored in a snapshot', async () => {
```
Comment (trioletas) [inline src/graphql/tests/joke.test.ts]: too verbose, reduce to `should match snapshot`

File: src/graphql/tests/__snapshots__/joke.test.ts.snap [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`joke schema should answer with a joke as stored in a snapshot 1`] = `
+Object {
+  "joke": null,
```
Comment (trioletas) [inline src/graphql/tests/__snapshots__/joke.test.ts.snap]: would expect a full payload here, see the details below

File: src/services/__service_mocks/api_jokes_random_limitTo_nerdy.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import * as nock from 'nock';
+
+const response = {
+  type: 'success',
+  'value': {
```
Comment (trioletas) [inline src/services/__service_mocks/api_jokes_random_limitTo_nerdy.ts]: `'value'` => `value`

**Comment by aeldar** [comment]
1. HTTP mocks (`nock`) moved inside integration tests. Despite the fact, that we aren't suppose to know anything about services on this level, it definitely looks good, and easier to understand what is going on.
2. Integration tests related directory names refactored for the sake of consistency. At first, they differed because I didn't want to add _ignore_ patterns to config. But that is no problem.
3. Also, example service test `sum-test.ts` renamed into `sum.test.ts`, if no one's mind. And added `.test` suffix to regex pattern for tests inside jest config, as there are not only tests, but also set-up files inside `__tests__`.
4. Jest configurations are DRY-ed. A `jest.default.config.js` is created in the root directory. There is also a nuance with it - we are not supposed to use jest template variables in it (like `<RootDir>`), because they will be wrong on deeper level configuration. It seems like a lot of jest configs in the root for now. Hope it is ok.

File: src/graphql/__tests__/joke.test.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    .query({ limitTo: '[nerdy]' })
+    .reply(200, response);
+});
+
+describe('query.joke', () => {
+  it('should match snapshot', async () => {
+    const query = `
+      query Q {
+        joke {
+          text
```
Comment (trioletas) [inline src/graphql/__tests__/joke.test.ts]: Add category

File: src/services/__service_mocks/api_jokes_random_limitTo_nerdy.ts (hunk 6 above)
Comment (aeldar) [inline src/services/__service_mocks/api_jokes_random_limitTo_nerdy.ts]: I am used to use linter for such a mistakes. Probably linter rules should be improved.

**Comment by trioletas** [comment]
Thanks a loot, good to merge.

**Pull request merged**